src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -0,0 +1,502 @@
+    lower(param).filterNot(isLess(_, butNot))
+    
+  def exclusiveUpper(param: PolyParam, butNot: PolyParam): List[PolyParam] =
+    upper(param).filterNot(isLess(butNot, _))
+  
+// ---------- Info related to PolyParams -------------------------------------------
+
+  def isLess(param1: PolyParam, param2: PolyParam): Boolean =
+    upper(param1).contains(param2)
+
+  def nonParamBounds(param: PolyParam): TypeBounds = 
+    entry(param).asInstanceOf[TypeBounds]
+    
+  def fullLowerBound(param: PolyParam)(implicit ctx: Context): Type =
+    (nonParamBounds(param).lo /: minLower(param))(_ | _)
+
+  def fullUpperBound(param: PolyParam)(implicit ctx: Context): Type = 
+    (nonParamBounds(param).hi /: minUpper(param))(_ & _)
+    
+  def fullBounds(param: PolyParam)(implicit ctx: Context): TypeBounds =
+    nonParamBounds(param).derivedTypeBounds(fullLowerBound(param), fullUpperBound(param))
+
+  def typeVarOfParam(param: PolyParam): Type = {
+    val entries = boundsMap(param.binder)
+    if (entries == null) NoType
+    else {
+      val tvar = typeVar(entries, param.paramNum)
+      if (tvar != null) tvar else NoType
+    }
+  }  
+
+// ---------- Adding PolyTypes --------------------------------------------------
+   
+  /** The bound type `tp` without dependent parameters
+   *  NoType if type consists only of dependent parameters.
+   *  @param seenFromBelow   If true, `bound` is an upper bound, else a lower bound.
+   */
+  private def stripParams(tp: Type, paramBuf: mutable.ListBuffer[PolyParam], 
+      isUpper: Boolean)(implicit ctx: Context): Type = tp match {
+    case param: PolyParam if contains(param) => 
+      if (!paramBuf.contains(param)) paramBuf += param
+      NoType
+    case tp: AndOrType if isUpper == tp.isAnd =>
+      val tp1 = stripParams(tp.tp1, paramBuf, isUpper)
+      val tp2 = stripParams(tp.tp2, paramBuf, isUpper)
+      if (tp1.exists)
+        if (tp2.exists) tp.derivedAndOrType(tp1, tp2)
+        else tp1
+      else tp2
+    case _ =>
+      tp
+  } 
+  
+  /** The bound type `tp` without dependent parameters.
+   *  A top or bottom type if type consists only of dependent parameters.
+   *  @param seenFromBelow   If true, `bound` is an upper bound, else a lower bound.
+   */
+  private def normalizedType(tp: Type, paramBuf: mutable.ListBuffer[PolyParam], 
+      isUpper: Boolean)(implicit ctx: Context): Type = 
+    stripParams(tp, paramBuf, isUpper)
+      .orElse(if (isUpper) defn.AnyType else defn.NothingType)
+  
+  def add(poly: PolyType, tvars: List[TypeVar])(implicit ctx: Context): This = {
+    assert(!contains(poly))
+    val nparams = poly.paramNames.length
+    val entries1 = new Array[Type](nparams * 2)
+    poly.paramBounds.copyToArray(entries1, 0)
+    tvars.copyToArray(entries1, nparams)
+    newConstraint(boundsMap.updated(poly, entries1), lowerMap, upperMap).init(poly)
+  }
+    
+  private def init(poly: PolyType)(implicit ctx: Context): This = {
```

Comment:
`add` is commented in Constraint.scala (as are all other public methods of OrderingConstraint).